myst_nb/cache.py

```diff
@@ -0,0 +1,192 @@
+                source_path = nb
+            else:
+                source_path = env.env.doc2path(nb)
+
+            has_outputs = read_nb_output_cells(source_path, has_outputs, do_run)
+
+            if not has_outputs:
+                if pk_list is None:
+                    pk_list = []
+                stage_record = cache_base.stage_notebook_file(source_path)
+                pk_list.append(stage_record.pk)
+            else:
+                if has_outputs:  # directly cache the already executed notebooks
+                    cache_base.cache_notebook_file(source_path, overwrite=True)
+                logger.error(
+                    f"Will not run notebook with pre-populated outputs/no output cells, will directly cache: {source_path}"  # noqa:E501
+                )
+
+        execute_staged_nb(
+            cache_base, pk_list
+        )  # can leverage parallel execution implemented in jupyter-cache here
+
+
+def add_notebook_outputs(env, ntbk, file_path=None):
+    """
+    Add outputs to a NotebookNode by pulling from cache.
+
+    Function to get the database instance. Get the cached output of the notebook
+    and merge it with the original notebook. If there is no cached output,
+    checks if there was error during execution, then saves the traceback to a log file.
+    """
+    # If we have a jupyter_cache, see if there's a cache for this notebook
+    file_path = file_path or env.doc2path(env.docname)
+    dest_path = env.app.outdir
+    reports_dir = dest_path + "/reports"
+    path_cache = False
+
+    if env.config["jupyter_cache"]:
+        path_cache = env.path_cache
+
+    if path_cache:
+        cache_base = get_cache(path_cache)
+        db = cache_base.db
+        cache_record = None
+        r_file_path = _relative_file_path(file_path)
+        try:
+            cache_list = NbCacheRecord.records_from_uri(file_path, db)
+            if len(cache_list):
+                latest = None
+                for item in cache_list:
+                    if latest is None or (latest < item.created):
+                        latest = item.created
+                        latest_record = item
+                cache_record = latest_record
+        except KeyError:
+            cache_record = None
+            logger.error(
+                (
+                    f"Couldn't find cache key for notebook file {r_file_path}. "
+                    "Outputs will not be inserted"
+                )
+            )
+        if cache_record:
+            _, ntbk = cache_base.merge_match_into_notebook(ntbk)
+        else:
+            try:
+                stage_record = cache_base.get_staged_record(file_path)
+            except KeyError:
+                stage_record = None
+            if stage_record and stage_record.traceback:
+                # save the traceback to a log file
+                ensuredir(reports_dir)
+                file_name = r_file_path[
+                    r_file_path.rfind("/") + 1 : r_file_path.rfind(".")
```

Comment:
should these be `os.sep`s? (maybe pathlib is smart about figuring out paths with `/` on non-*nix systems?)

Reply:
have used `os.path.splitext(r_file_path.name)[0]` instead of this now